CONTRIBUTING.md

```diff
@@ -23,19 +23,25 @@ Things that should go into your PR description:
  * A changelog entry in the `Notes` section (see below)
  * References to any bugs fixed by the change (in GitHub's `Fixes` notation)
  * Describe the why and what is changing in the PR description so it's easy for
-   onlookers and reviewers to onboard and context switch.
+   onlookers and reviewers to onboard and context switch. This information is
+   also helpful when we come to look at this in 6 months and ask "why did we do
+   it like that?" we have a chance of finding out.
+      * Why is the change necessary? What problem does it solve? Why didn't it
+        work before? Why does it work now? What use cases does it unlock?
```

Comment:
Our best issues at the moment only explain the problem and most of the time, just the symptoms. The potential solutions section is rarely filled out and the solutions explained there are not fleshed out (just nuggets of an idea). We really need to commit to keeping the issue description as the single source of truth, see https://about.gitlab.com/handbook/communication/#issues. But this wasn't what I was hearing during a web chapter sync discussion. There was a bias to not edit peoples issues where the reasoning was because they are an append-only log and editing peoples messages is some variation of bad-practice/evil. Of course it doesn't always make sense but there is plenty of reasonable scenarios to improve things. We also can't ask external contributors to keep an issue description from a different author up to date though and reading through the entire history of an issue is cumbersome just to get the context needed.

The context for all of this should be in the PR description though. Sometimes, it would be better to point to or quote from the issue where it's already better explained and summarized, that makes sense. Plus, [not every PR starts with an issue](https://about.gitlab.com/handbook/communication/#everything-starts-with-a-merge-request) so it's good to make sure the information is available in this case.

In order to not always require reading the context, the PR title should be a one-line summary itself and the first part of the description can fill that in with an extra sentence before diving into the details below that.


---

Let's take https://github.com/vector-im/element-web/issues/16983 as one of the good issue examples which explains the symptoms, what the problem probably is and a potential solution. But from the PR https://github.com/matrix-org/matrix-react-sdk/pull/8563, we don't know if that was the actual problem or the solution used. It all of sudden makes a lot more sense, once we get context: https://github.com/matrix-org/matrix-react-sdk/pull/8563#discussion_r870687570

Reply:
The gitlab handbook has been consistently at a different bar than where we're at or want to be fwiw, so would hesitate to use it as a comparison tool here. 

Issue context should always be in the issue, not the PR. It's true that contributors can't edit the issue and that our issues kinda suck, but the point is that we fix that as best we can rather than degrade the review process. This is because the review process is generally more critical than the triage/investigation process, as it literally halts progress if not done fast enough.

A proposal would be to have the team monitor issues more closely to ensure they're still serving a useful purpose in the backlog, thus encouraging a higher quality of issue for new reports. We might even change our entire intake process or templates to accommodate improved issue quality. Building out such a process keeps issue context largely out of the PR, making it easier and faster to review a given change.

I also personally disagree that we should be able to start with a PR - nearly everything should be attributed to a backlog item somewhere. This is a discussion for a different audience, however.

For clarity: I'm asking that we keep problem details out of the PR, not detail overall. Our PRs should be descriptive enough where they can stand on their own, though when describing a bug it's more important to keep all the details about that bug on the ticket.

---

On your example: the expectation has generally been that people familiarize themselves with the code. That's not to say I agree with the approach that was taken in that specific instance, however it's very common to fix an issue without context because the context is in the diff (when interpreted). We do know that it fixed the issue too, as it said it closed the issue, and did. 

The example does demonstrate a great way to introduce issue details into a PR though: the PR's title is (nearly) phrased for an end user to understand what is going on and doubles as an explanation of the problem. Without looking at the issue, I know from the PR title alone that the referenced issue will be about how history is broken after join because we're not considering some decryption keys. I could confirm this in the issue easily if I wanted to (I might if I were a bit newer to the team), though the code itself reinforces the assumptions about the issue too (and in fewer words).

Reply:
> A proposal would be to have the team monitor issues more closely to ensure they're still serving a useful purpose in the backlog, thus encouraging a higher quality of issue for new reports. We might even change our entire intake process or templates to accommodate improved issue quality. Building out such a process keeps issue context largely out of the PR, making it easier and faster to review a given change.

👍 Improving issue quality is great. It's a bit parallel and ⚖ where one can make up for the other in some aspects (see different phases of work paragraph below).

> For clarity: I'm asking that we keep problem details out of the PR, not detail overall. Our PRs should be descriptive enough where they can stand on their own, though when describing a bug it's more important to keep all the details about that bug on the ticket.

I think we should at least note that these details should be on the issue. Like a section that says "Make sure the issue is up to date with these things..."

I think they are addressing two different phases of work though. When it comes to the issue, the main thing we're trying to understand is the symptoms, how to reproduce it, and expectations on how it should behave. It's rare and special when you can think of the exact problem piece of code before diving in. Then for the PR, we find out what is actually going wrong, where did it regress to find context on why it was done that way initially, and the fix. This is why "Why didn't it work before? Why does it work now?" is most relevant to the PR. 


> On your example: the expectation has generally been that people familiarize themselves with the code. That's not to say I agree with the approach that was taken in that specific instance, however it's very common to fix an issue without context because the context is in the diff (when interpreted).

I think this is mostly where we are not seeing eye to eye. While we strive for code and comments that is clear and self-explanatory, this is not reality (otherwise we would be able to spot every bug before shipping). Going back to fixes without this information after months, years makes it soo much harder understand. I could never get the information that was provided in https://github.com/matrix-org/matrix-react-sdk/pull/8563#discussion_r870687570 from the diff.

> The example does demonstrate a great way to introduce issue details into a PR though: the PR's title is (nearly) phrased for an end user to understand what is going on and doubles as an explanation of the problem. Without looking at the issue, I know from the PR title alone that the referenced issue will be about how history is broken after join because we're not considering some decryption keys. I could confirm this in the issue easily if I wanted to (I might if I were a bit newer to the team), though the code itself reinforces the assumptions about the issue too (and in fewer words).

This still only explains the "what", not the "why". Which is what I'm trying to propose we be better at in this PR.

To call out a recent PR that has done this well, see https://github.com/matrix-org/matrix-ios-sdk/pull/1483

Reply:
It sort of feels like we're saying almost the same thing, but still different. I find that iOS SDK PR to be way too verbose to even start engaging with it as a reviewer, but presumably that's what works for that team. 

We should absolutely be describing what the problem is, but not copy the issue details into the PR. If it's any clearer, I'm trying to *remove* these two questions:
> Why is the change necessary? What problem does it solve?

These are both answered by a "Fixes https://..." reference. The other 3 questions I would also argue shouldn't be in the PR description necessarily, but it's a detail at code review time that we can work out and decide if it needs to be in the code itself.

Reply:
> I'm trying to *remove* these two questions:
> > Why is the change necessary? What problem does it solve?

Removed so we can move this forward ⏩

Technically this can work and better issues ftw 👍